fuse_optimizers/test/test_fixed_lag_ignition.cpp

```diff
@@ -0,0 +1,124 @@
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <fuse_models/SetPose.h>
+#include <geometry_msgs/PoseWithCovarianceStamped.h>
+#include <nav_msgs/Odometry.h>
+#include <ros/ros.h>
+
+#include <gtest/gtest.h>
+
+
+TEST(FixedLagIgnition, SetInitialState)
+{
+  auto node_handle = ros::NodeHandle();
+  auto pose_publisher = node_handle.advertise<geometry_msgs::PoseWithCovarianceStamped>("/pose", 1);
+
+  // Wait for the optimizer to be ready
+  ASSERT_TRUE(ros::service::waitForService("/fixed_lag/reset", ros::Duration(1.0)));
+  ASSERT_TRUE(ros::service::waitForService("/fixed_lag/set_pose", ros::Duration(1.0)));
+
+  // Set the initial pose to something near zero
+  fuse_models::SetPose::Request req;
+  req.pose.header.frame_id = "map";
+  req.pose.header.stamp = ros::Time(1, 0);
+  req.pose.pose.pose.position.x = 100.1;
+  req.pose.pose.pose.position.y = 100.2;
+  req.pose.pose.pose.position.z = 0.0;
+  req.pose.pose.pose.orientation.x = 0.0;
+  req.pose.pose.pose.orientation.y = 0.0;
+  req.pose.pose.pose.orientation.z = 0.0;
+  req.pose.pose.pose.orientation.w = 1.0;
+  req.pose.pose.covariance[0] = 1.0;
+  req.pose.pose.covariance[7] = 1.0;
+  req.pose.pose.covariance[35] = 1.0;
+  fuse_models::SetPose::Response res;
+  ros::service::call("/fixed_lag/set_pose", req, res);
+
+  // I need to wait for the subscriber to be ready. I don't know of a better way.
+  // There is no ros::topic::waitForSubscriber() method.
+  ros::Duration(0.5).sleep();
+
+  // Publish a relative pose
+  auto pose_msg1 = geometry_msgs::PoseWithCovarianceStamped();
+  pose_msg1.header.stamp = ros::Time(2, 0);
+  pose_msg1.header.frame_id = "base_link";
+  pose_msg1.pose.pose.position.x = 0.0;
+  pose_msg1.pose.pose.position.y = 0.0;
+  pose_msg1.pose.pose.position.z = 0.0;
+  pose_msg1.pose.pose.orientation.x = 0.0;
+  pose_msg1.pose.pose.orientation.y = 0.0;
+  pose_msg1.pose.pose.orientation.z = 0.0;
+  pose_msg1.pose.pose.orientation.w = 0.0;
+  pose_msg1.pose.covariance[0] = 1.0;
+  pose_msg1.pose.covariance[7] = 1.0;
+  pose_msg1.pose.covariance[35] = 1.0;
+  pose_publisher.publish(pose_msg1);
+
+  auto pose_msg2 = geometry_msgs::PoseWithCovarianceStamped();
+  pose_msg2.header.stamp = ros::Time(3, 0);
+  pose_msg2.header.frame_id = "base_link";
+  pose_msg2.pose.pose.position.x = 10.0;
+  pose_msg2.pose.pose.position.y = 20.0;
+  pose_msg2.pose.pose.position.z = 0.0;
+  pose_msg2.pose.pose.orientation.x = 0.0;
+  pose_msg2.pose.pose.orientation.y = 0.0;
+  pose_msg2.pose.pose.orientation.z = 0.5000;
+  pose_msg2.pose.pose.orientation.w = 0.8660;
+  pose_msg2.pose.covariance[0] = 1.0;
+  pose_msg2.pose.covariance[7] = 1.0;
+  pose_msg2.pose.covariance[35] = 1.0;
+  pose_publisher.publish(pose_msg2);
+
+  // Wait for the optimizer to publish the first pose
+  auto odom_msg = ros::topic::waitForMessage<nav_msgs::Odometry>("/odom", ros::Duration(1.0));
```

Comment:
Same question for 1 second timeout, is it enough or should wait forever?

Reply:
The fixed-lag smoother is configured to update every 1.0s seconds. So maybe adding a bit more time would be safer. But if I don't receive a message, that should be considered a failure case.

Reply:
Ideally ROS would not return from the `node_handle.subscribe()` call until the subscriber was ready. But there's nothing I can do about that. 🙂

There is a complex set of interactions between the "set_pose" service call and the initialization of the subscriber. That makes waiting for a message awkward. Even if I could do as you suggest, it would be self defeating. Waiting to publish a message until a message has been received is somewhat self-defeating.